evm/p2p/lightchain.py

```diff
@@ -1,69 +1,270 @@
+                    kademlia.Address("51.15.218.125", 30303, 30303)),
+                kademlia.Node(
+                    keys.PublicKey(decode_hex("e80276aabb7682a4a659f4341c1199de79d91a2e500a6ee9bed16ed4ce927ba8d32ba5dea357739ffdf2c5bcc848d3064bb6f149f0b4249c1f7e53f8bf02bfc8")),  # noqa: E501
+                    kademlia.Address("51.15.39.57", 30303, 30303)),
+                kademlia.Node(
+                    keys.PublicKey(decode_hex("584c0db89b00719e9e7b1b5c32a4a8942f379f4d5d66bb69f9c7fa97fa42f64974e7b057b35eb5a63fd7973af063f9a1d32d8c60dbb4854c64cb8ab385470258")),  # noqa: E501
+                    kademlia.Address("51.15.35.2", 30303, 30303)),
+                kademlia.Node(
+                    keys.PublicKey(decode_hex("d40871fc3e11b2649700978e06acd68a24af54e603d4333faecb70926ca7df93baa0b7bf4e927fcad9a7c1c07f9b325b22f6d1730e728314d0e4e6523e5cebc2")),  # noqa: E501
+                    kademlia.Address("51.15.132.235", 30303, 30303)),
+                kademlia.Node(
+                    keys.PublicKey(decode_hex("482484b9198530ee2e00db89791823244ca41dcd372242e2e1297dd06f6d8dd357603960c5ad9cc8dc15fcdf0e4edd06b7ad7db590e67a0b54f798c26581ebd7")),  # noqa: E501
+                    kademlia.Address("51.15.75.138", 30303, 30303)),
+            ]
+        else:
+            raise ValueError("Unknown network_id: %s", self.network_id)
 
 
 class LightChain(Chain):
-    on_demand_data_backend_class = None
+    # FIXME: Should be passed in by callers
+    privkey = ecies.generate_privkey()
+    # TODO:
+    # 1. Implement a queue of requests and a distributor which picks items from that queue and
+    # sends them to one of our peers, ensuring the selected peer has the info we want,
+    # retrying on timeouts and respecting the flow control rules.
 
-    def __init__(self, chaindb, header=None):
-        super(LightChain, self).__init__(chaindb, header=header)
-        self.on_demand_data_backend = self.on_demand_data_backend_class(self.chaindb)
+    def __init__(self, chaindb):
+        super(LightChain, self).__init__(chaindb)
+        self.peer_pool = PeerPool(chaindb, self.network_id, self.privkey)
+
+    async def run(self):
+        await self.peer_pool.run()
 
     async def stop(self):
-        await self.on_demand_data_backend.stop()
+        self.logger.info("Stopping ...")
+        await self.peer_pool.stop()
 
     async def get_canonical_block_by_number(self, block_number):
         try:
             block_hash = self.chaindb.lookup_block_hash(block_number)
         except KeyError:
-            raise BlockNotFound("No block with number {} found on canonical chain".format(
-                block_number))
+            raise BlockNotFound(
+                "No block with number %s found on local chain", block_number)
         return await self.get_block_by_hash(block_hash)
 
+    @alru_cache(maxsize=1024)
     async def get_block_by_hash(self, block_hash):
+        peer = await self.peer_pool.get_best_peer()
```

Comment:
As I read it, if we issue a bunch of requests and all of our peers are the same height, we would hammer the peer at the beginning of the list with all of the requests.

Am I reading it right? If so, maybe it's worth adding a tracking issue to spread the load across the peers.

Reply:
That suggests to me that the strategy for choosing the *best* peer should pick randomly from the set of peers who have the highest block height.

Reply:
Yeah, makes sense. Since at this point it is used only for on-demand requests (and we only fetch blocks on demand), would it be ok if I add a TODO comment to address that later?

Reply:
> would it be ok if I add a TODO comment to address that later?

Yeah, that's fine.  Maybe even write up an issue for this and while I don't think it's an easy-pickings level thing, it's something that someone could come in and implement it without much base knowledge of the codebase.